### ISSUE TITLE
Authentication provided by OneDrive rejected afterwards

### ISSUE THREAD

**Issue by Poshi** [abraunegg/onedrive#1683]
### Bug Report Details ###
I'm trying to connect to my OneDrive unit, but I cannot get past the authentication point.

**Describe the bug**
I run `onedrive` to get the authentication URI. I cut&paste it into my browser. I authenticate myself and give permissions. I copy the URI back to the terminal. The program complains and gaves me a stack trace.

Terminal trace (UUIDs redacted):
```
$ onedrive --verbose --verbose
[DEBUG] homePath: HOME environment variable set
[DEBUG] homePath: /home/username
[DEBUG] configDirBase: WARNING - no XDG_CONFIG_HOME environment variable set
[DEBUG] configDirBase: /home/username/.config
[DEBUG] configDirName: Configuring application to use default config path
Using Config Dir: /home/username/.config/onedrive
No config file found, using application defaults
[DEBUG] sync_dir: Getting syncDir from config value sync_dir
[DEBUG] sync_dir: A '~' was found in configured sync_dir, automatically expanding as SHELL and USER environment variable is set
[DEBUG] syncDir: /home/username/OneDrive
Initializing the OneDrive API ...
[DEBUG] Downgrading all HTTP operations to HTTP/1.1 by default
Authorize this app visiting:

https://login.microsoftonline.com/common/oauth2/v2.0/authorize?client_id=xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx&scope=Files.ReadWrite%20Files.ReadWrite.all%20Sites.ReadWrite.All%20offline_access&response_type=code&redirect_uri=https://login.microsoftonline.com/common/oauth2/nativeclient

Enter the response uri: https://login.microsoftonline.com/common/oauth2/nativeclient?code=M.R3_BL2.xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
OneDrive returned a 'HTTP 400 - Bad Request' - gracefully handling error
onedrive.OneDriveException@src/onedrive.d(874): HTTP request returned status code 400 (Bad Request)
{
    "correlation_id": "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx",
    "error": "invalid_grant",
    "error_codes": [
        70000
    ],
    "error_description": "AADSTS70000: The provided value for the 'code' parameter is not valid.\r\nTrace ID: xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx\r\nCorrelation ID: xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx\r\nTimestamp: 2021-11-09 14:58:39Z",
    "error_uri": "https:\/\/login.microsoftonline.com\/error?code=70000",
    "timestamp": "2021-11-09 14:58:39Z",
    "trace_id": "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx"
}
----------------
??:? [0x55f76fca99d9]
??:? [0x55f76fca8cb5]
??:? [0x55f76fca9b65]
??:? [0x55f76fca7f58]
??:? [0x55f76fca78c5]
??:? [0x55f76fcb57a8]
??:? void rt.dmain2._d_run_main2(char[][], ulong, extern (C) int function(char[][])*).runAll() [0x7f87855309db]
??:? _d_run_main2 [0x7f87855307ee]
??:? _d_run_main [0x7f878553065d]
??:? __libc_start_main [0x7f878511c0b2]
??:? [0x55f76fc805ed]
```

**Application and Operating System Details:**
*   Provide your OS & version (CentOS 6.x, Ubuntu 18.x etc) and the output of: `uname -a`
Ubuntu 20.04.3 LTS
Linux computer 5.4.0-90-generic #101-Ubuntu SMP Fri Oct 15 20:00:55 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
*   Are you using a headless system (no gui) or with a gui installed?
GUI installed, but trying to run the onedrive program as a daemon.
*   OneDrive Account Type
Personal
*   Did you build from source or install from a package?
Built from source
*   If you installed from source, what is your DMD or LDC compiler version: `dmd --version` or `ldmd2 --version`
DMD64 D Compiler v2.098.0
Copyright (C) 1999-2021 by The D Language Foundation, All Rights Reserved written by Walter Bright
*   OneDrive Application Version: Output of `onedrive --version`
onedrive v2.3.13-1build2
*   OneDrive Application Configuration: Output of `onedrive --display-config`
```
onedrive version                    = v2.3.13-1build2
Config path                         = /home/username/.config/onedrive
Config file found in config path    = false
Config option 'check_nosync'        = false
Config option 'sync_dir'            = /home/username/OneDrive
Config option 'skip_dir'            =
Config option 'skip_file'           = ~*|.~*|*.tmp
Config option 'skip_dotfiles'       = false
Config option 'skip_symlinks'       = false
Config option 'monitor_interval'    = 45
Config option 'min_notify_changes'  = 5
Config option 'log_dir'             = /var/log/onedrive/
Config option 'sync_root_files'     = false
Selective sync configured           = false
```
*   Provide the version of curl you are using: Output of `curl --version`
curl 7.68.0 (x86_64-pc-linux-gnu) libcurl/7.68.0 OpenSSL/1.1.1f zlib/1.2.11 brotli/1.0.7 libidn2/2.2.0 libpsl/0.21.0 (+libidn2/2.2.0) libssh/0.9.3/openssl/zlib nghttp2/1.40.0 librtmp/2.3
Release-Date: 2020-01-08
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp
Features: AsynchDNS brotli GSS-API HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL TLS-SRP UnixSockets
*   Is your configured 'sync_dir' a local directory or a network mount point?
Local directory
*   If *not* local, provide all the mountpoints in your system: Output of: `mount`
N/A
*   What partition format type does your configured 'sync_dir' reside on? Output of: `lsblk -f` 
`└─sdb4 ext4                 7d80f890-76d7-4cce-9eac-77583c3cf5da  233.4G    48% /home`
*   Explain your entire configuration setup - is the OneDrive folder shared with any other system, shared with any other platform at the same time, is the OneDrive account you use shared across multiple systems / platforms / Operating Systems and in use at the same time
One drive folder shared between two computers (home/work). No parallel accesses expected.

**To Reproduce**
Steps shown above.

**Complete Verbose Log Output**
See above

### Bug Report Checklist ###
*   [x] Detailed description
*   [x] Application and Operating System Details provided in full
*   [x] Reproduction steps (if applicable)
*   [x] Verbose Log Output from your error
*   [x] Debug Log generated and submitted

**Comment by abraunegg** [comment]
@Poshi 
This is because you have installed from Ubuntu default repositories which is not supported - thus you have installed an old version which was released in 2019. 

Please follow the correct installation process for Ubuntu platforms.

**Comment by Poshi** [comment]
Hummm...
You were wrong, at least in part.
I installed from source (git clone...), but had a non-official repository added from "ancient times" (pre-pandemic). It seems that the recently compiled version was interacting somehow with the installed version.
By chance, the repository got updated these last two days, and a `make uninstall` and an update of the repository leaved the one from the repository in working conditions.
Sorry for the bug report, I followed all the instructions and was genuinely convinced it was a bug. I was not aware of any software interaction.
Thanks!

**Comment by abraunegg** [comment]
@Poshi 
I may be wrong about your setup - your bug report lists `v2.3.13-1build2` and the 'crash' is because that version has the authentication bug because Microsoft changed their authentication.

If you are using the Ubuntu packages - you need to be critically aware of these points: 
* The 'onedrive' client will automatically startup post 'authentication' without any further actions
* The client will segfault | core-dump when exiting

See https://github.com/abraunegg/onedrive/blob/master/docs/ubuntu-package-install.md#known-issues-with-installing-from-the-above-packages for further details

**Comment by Poshi** [comment]
Yes, you are right at those points. But I can swear that I installed from the HEAD of the github repository. So it had to be some interaction with the package that I was not aware it was already installed. The repository used not the Ubuntu one, it is `http://ppa.launchpad.net/yann1ck/onedrive/ubuntu`.
In any case, the issue has been solved and I managed to authenticate, thanks to your pointer to an outdated version.
